### DISCUSSION TITLE
Text examples: Custom tokenizer

### DISCUSSION THREAD

**Discussion by nn09** [shap/shap#2171] [Q&A] [answered]
Hello,

I am trying to define a custom tokenizer for text data, as described at https://shap.readthedocs.io/en/latest/example_notebooks/text_examples/sentiment_analysis/Using%20custom%20functions%20and%20tokenizers.html

More specifically, I re-used:
method = "custom tokenizer"

if method == "transformers tokenizer":
    explainer = shap.Explainer(f, tokenizer, output_names=labels)

elif method == "default masker":
    masker = shap.maskers.Text(r"\W") # this will create a basic whitespace tokenizer
    explainer = shap.Explainer(f, masker, output_names=labels)

elif method == "custom tokenizer":
    import re

    def custom_tokenizer(s, return_offsets_mapping=True):
        """ Custom tokenizers conform to a subset of the transformers API.
        """
        pos = 0
        offset_ranges = []
        input_ids = []
        for m in re.finditer(r"\W", s):
            start, end = m.span(0)
            offset_ranges.append((pos, start))
            input_ids.append(s[pos:start])
            pos = end
        if pos != len(s):
            offset_ranges.append((pos, len(s)))
            input_ids.append(s[pos:])
        out = {}
        out["input_ids"] = input_ids
        if return_offsets_mapping:
            out["offset_mapping"] = offset_ranges
        return out

    masker = shap.maskers.Text(custom_tokenizer)
    explainer = shap.Explainer(f, masker, output_names=labels)

However, I obtain the following error message:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-1d4129d9e966> in <module>
     34         return out
     35 
---> 36     masker = shap.maskers.Text(custom_tokenizer)
     37     explainer = shap.Explainer(f, masker, output_names=labels)

~/miniconda3/envs/py39/lib/python3.9/site-packages/shap/maskers/_text.py in __init__(self, tokenizer, mask_token, collapse_mask_token, output_type)
     21         self.collapse_mask_token = collapse_mask_token
     22         self.input_mask_token = mask_token
---> 23         parsed_tokenizer_dict = parse_prefix_suffix_for_tokenizer(tokenizer)
     24 
     25         self.keep_prefix = parsed_tokenizer_dict['keep_prefix']

~/miniconda3/envs/py39/lib/python3.9/site-packages/shap/utils/transformers.py in parse_prefix_suffix_for_tokenizer(tokenizer)
     89     used to slice tokens belonging to sentence after passing through tokenizer.encode().
     90     """
---> 91     null_tokens = tokenizer.encode("")
     92     keep_prefix, keep_suffix, prefix_strlen, suffix_strlen = None, None, None, None
     93 

AttributeError: 'function' object has no attribute 'encode'

Has anyone encountered similar issues, and how did you overcome it?

**Comment by YueyuanKang** [comment, marked as answer]
I have also found this problem. You can download the shap file in the github. And exchange the shap file in your library file with this one. It will work. Have a good day.

**Reply by nn09** [reply]
Thanks a lot!

**Comment by YueyuanKang** [comment]
The version in the Github will not have this problem. But version 0.39.0 will have this problem. And I am also working on this example. I have got the shap values and feature names matrix. The problem I face is that I can not get the type of the matrix which the shap.plot want. If you are also working on this example, I think we can discuss it.

**Reply by nn09** [reply]
Yes, certainly. Can you elaborate on the problem? For example, can you post some lines to replicate the issue?

**Reply by YueyuanKang** [reply]
I am not familiar with GitHub. If you want to solve it. You can email me with kangyueyuan2019@163.com. I have solved this problem. And I can visualize the result with the function shap.plots.text.
![image](https://user-images.githubusercontent.com/77316342/134303306-afb64671-e5da-42ba-ab5a-909c50d44815.png)

**Comment by mosh98** [comment]
Has someone been able to use custom tokenizer with roberta models/tokenizer. My text visualization looks super ugly

**Reply by dk25021999** [reply]
Did you find solution with other tokenizers?

**Reply by mosh98** [reply]
nep
